### PR TITLE
Create internal rust_binary rule instead of using transitions

### PR DESCRIPTION
**Pull request by dfreese**
This seems like a cleaner way to break the dependency that rust_binary
has on the process wrapper.  The common args are consumed and modified
to set the process wrapper to None, which is detected and performs the
appropriate switch.  The transitions and settings can then be removed.

### CODE REVIEW

**Comment by ddeville** [comment]
I can confirm that this fixes some cycles I was seeing when querying the bazel dependency graph after #1159.

File: rust/private/rustc.bzl [hunk 1]
```diff
@@ -922,8 +922,7 @@ def rustc_compile_action(
             dsym_folder = ctx.actions.declare_directory(crate_info.output.basename + ".dSYM")
             action_outputs.append(dsym_folder)
 
-    # This uses startswith as on windows the basename will be process_wrapper_fake.exe.
-    if not ctx.executable._process_wrapper.basename.startswith("process_wrapper_fake"):
+    if ctx.executable._process_wrapper != None:
```
Comment (UebelAndre) [inline rust/private/rustc.bzl:925]: Is this meaningfully different from:
```suggestion
    if ctx.executable._process_wrapper:
```

File: rust/private/rust.bzl [hunk 2]
```diff
@@ -928,6 +928,33 @@ rust_binary = rule(
...
+        executable = True,
+        allow_single_file = True,
+        cfg = "exec",
+    )
+    return new_attr
+
+# Provides an internal rust_binary to use that we can use to build the process
+# wrapper, this breaks the dependency of rust_binary on the process wrapper by
+# setting it to None, which the functions in rustc detect and build accordingly.
+rust_binary_for_process_wrapper = rule(
```
Comment (UebelAndre) [inline rust/private/rust.bzl:944]: small nit
```suggestion
rust_binary_without_process_wrapper = rule(
```

**Review by UebelAndre — CHANGES_REQUESTED** [review]
One question but otherwise looks good!

**Comment by dfreese** [comment]
All of those changes look fine to me.  I'll be able to get back to this later Sunday night, so feel free to update and merge. I think you should have permission.

**Review by UebelAndre — APPROVED** [review]
Thanks for working on this!

**Comment by hlopko** [comment]
Thanks! Indeed much simpler!

**Pull request merged**